Source/objects.cpp

```diff
@@ -2350,67 +2350,57 @@ void RedoPlayerVision()
 	}
 }
 
-void OperateL1RDoor(int pnum, int oi, unsigned char sendflag)
+void OperateL1RDoor(int pnum, int oi, BOOL sendflag)
 {
-	int v3;     // esi
-	int v4;     // eax
-	int v5;     // ebx
-	int v6;     // edi
-	int v7;     // ST04_4
-	int v8;     // [esp+Ch] [ebp-Ch]
-	int v9;     // [esp+10h] [ebp-8h]
-	int param1; // [esp+14h] [ebp-4h]
+	int xp, yp, pn;
 
-	v3 = oi;
-	param1 = oi;
-	v9 = pnum;
-	v4 = object[oi]._oVar4;
-	if (v4 != 2) {
-		v5 = object[v3]._ox;
-		v6 = object[v3]._oy;
-		if (v4) {
-			if (!deltaload)
-				PlaySfxLoc(IS_DOORCLOS, v5, object[v3]._oy);
-			v8 = v6 + 112 * v5;
-			if (dDead[0][v8] != 0 || dMonster[0][v8] != 0 || dItem[0][v8] != 0) {
-				object[v3]._oVar4 = 2;
-				return;
-			}
-			if (v9 == myplr && sendflag)
-				NetSendCmdParam1(TRUE, CMD_CLOSEDOOR, param1);
-			v7 = object[v3]._oVar1;
-			object[v3]._oVar4 = 0;
-			object[v3]._oSelFlag = 3;
-			ObjSetMicro(v5, v6, v7);
-			if (object[v3]._oVar2 == 50) {
-				if (dPiece[-1][v8] == 396) /* check *(_DWORD *)&dflags[28][4 * v8 + 32] == 396 ) */
-					ObjSetMicro(v5 - 1, v6, 411);
-				else
-					ObjSetMicro(v5 - 1, v6, 50);
+	if (object[oi]._oVar4 == 2) {
+		if (!deltaload)
+			PlaySfxLoc(IS_DOORCLOS, object[oi]._ox, object[oi]._oy);
+		return;
+	}
+
+	xp = object[oi]._ox;
+	yp = object[oi]._oy;
+	if (object[oi]._oVar4 == 0) {
+		if (pnum == myplr && sendflag)
+			NetSendCmdParam1(TRUE, CMD_OPENDOOR, oi);
+		if (!deltaload)
+			PlaySfxLoc(IS_DOOROPEN, object[oi]._ox, object[oi]._oy);
+		ObjSetMicro(xp, yp, 395);
+		dArch[xp][yp] = 8;
+		objects_set_door_piece(xp, yp - 1);
+		object[oi]._oAnimFrame += 2;
+		object[oi]._oPreFlag = TRUE;
+		DoorSet(oi, xp - 1, yp);
+		object[oi]._oVar4 = 1;
+		object[oi]._oSelFlag = 2;
+		RedoPlayerVision();
+	} else {
+		if (!deltaload)
+			PlaySfxLoc(IS_DOORCLOS, xp, object[oi]._oy);
+		if (((dDead[xp][yp] != 0 ? 0 : 1) & (dMonster[xp][yp] != 0 ? 0 : 1) & (dItem[xp][yp] != 0 ? 0 : 1)) != 0) {

```

Comment:
We should try to flip the boolean logic here to make it more readable if possible. The line essentially reads, if *all of them are zero, is _not_ true*, in other words if *one of them is not zero*.

A few suggestions (only consider them if they too are bin exact).

```suggestion
		if (((dDead[xp][yp] == 0) & (dMonster[xp][yp] == 0) & (dItem[xp][yp] == 0)) != 0) {
```

```suggestion
		if (!((dDead[xp][yp] == 0) & (dMonster[xp][yp] == 0) & (dItem[xp][yp] == 0))) {
```

```suggestion
		if ((dDead[xp][yp] != 0) || (dMonster[xp][yp] != 0) || (dItem[xp][yp] != 0)) {
```

Another variation of the above one using `|` rather than `||` to not use boolean circuit shortcuts (in the hope that it could affect bin accuracy).
```suggestion
		if ((dDead[xp][yp] != 0) | (dMonster[xp][yp] != 0) | (dItem[xp][yp] != 0)) {
```

Any other ideas how this line may have looked like in the original source?

Reply:
biwise `or` or anything logical would not work most likely as I have tested those.  First/second suggestions would probably work but I just prefer not to rely on result of logical expression being equal to `1` or `0` . As it would technically become `bool` in C++ and while it would probably work with bitwise operations, it still could rise a few eyebrows from static analyzers. 

All in all, I agree this code is not pretty.

Reply:
Interestingly enough even modern versions of Visual Studio are not eliminating branching in such expressions unlike other compilers:
https://gcc.godbolt.org/z/yV6nh_

Reply:
> biwise `or` or anything logical would not work most likely as I have tested those. First/second suggestions would probably work but I just prefer not to rely on result of logical expression being equal to `1` or `0` . As it would technically become `bool` in C++ and while it would probably work with bitwise operations, it still could rise a few eyebrows from static analyzers.
> 
> All in all, I agree this code is not pretty.

That's fair.

I'll merge this PR as is.

Reply:
> Interestingly enough even modern versions of Visual Studio are not eliminating branching in such expressions unlike other compilers:
> [godbolt](https://gcc.godbolt.org/z/yV6nh_)

Hmm, so they don't do circuit shortcutting. Good to know.